subprojects/execution/src/main/java/org/gradle/internal/execution/steps/BuildCacheStep.java

```diff
@@ -138,8 +138,14 @@ private AfterExecutionResult executeAndStoreInCache(CacheableWork cacheableWork,
         }
         AfterExecutionResult result = executeWithoutCache(cacheableWork.work, context);
         result.getExecution().ifSuccessfulOrElse(
-            executionResult -> result.getAfterExecutionState()
-                .ifPresent(afterExecutionState -> store(cacheableWork, cacheKey, afterExecutionState.getOutputFilesProducedByWork(), afterExecutionState.getOriginMetadata().getExecutionTime())),
+            executionResult -> {
+                if (executionResult.canStoreOutputsInCache()) {
+                    result.getAfterExecutionState()
+                        .ifPresent(afterExecutionState -> store(cacheableWork, cacheKey, afterExecutionState.getOutputFilesProducedByWork(), afterExecutionState.getOriginMetadata().getExecutionTime()));
+                } else {
+                    LOGGER.debug("Not storing result of {} in cache because reuse of its outputs was disabled", cacheableWork.getDisplayName());
+                }
+            },
```

Comment:
:x: We need to find a place to document how this new behavior works and why we support it. Maybe here?